File: src/0169_majority_element.js
/**
 * Boyer Moore Algorithm
 * Time O(N) | Space O(1)
 * https://leetcode.com/problems/majority-element
 * @param {number[]} nums
 * @return {number}
 */

var majorityElement = function (nums) {
    let res = nums[0];
    let count = 1;

    for (let i = 1; i < nums.length - 1; i++) {
        if (nums[i] === res) count++;
        else if (!--count) {
            res = nums[i + 1];
            count = 0;
        }
    }

    return res;
};

/**
 * HashMap
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/majority-element
 * @param {number[]} nums
 * @return {number}
 */

var majorityElement = function (nums) {
    let candidate = nums[0];
    let count = 0;
    for (let i = 0; i < nums.length; i++) {
        if (count === 0) {
            candidate = nums[i];
        }
        if (nums[i] === candidate) {
            count += 1;
        } else {
            count -= 1;
        }
    }
    return candidate;
};

module.exports = { majorityElement };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(majorityElement([3, 2, 3]), 3);
    assert.deepStrictEqual(majorityElement([2, 2, 1, 1, 1, 2, 2]), 2);
    assert.deepStrictEqual(majorityElement([1]), 1);
    assert.deepStrictEqual(majorityElement([6, 5, 5]), 5);
}

File: src/0169_majority_element.rs
struct Solution;

use std::collections::HashMap;

impl Solution {
    // Time Complexity: O(n)
    // Space Complexity: O(1)
    // Boyer - Moore Algorithm
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let (mut res, mut count) = (0, 0);

        for n in nums {
            if count == 0 {
                res = n;
            }
            count += if n == res { 1 } else { -1 };
        }
        res
    }

    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Hashmap
    pub fn majority_element_2(nums: Vec<i32>) -> i32 {
        let mut count = HashMap::new();

        let (mut res, mut max_count) = (0, 0);

        for num in nums {
            *count.entry(num).or_insert(0) += 1;
            res = if *count.get(&num).unwrap() > max_count {
                num
            } else {
                res
            };
            max_count = i32::max(*count.get(&num).unwrap(), max_count);
        }
        res
    }

    // Time Complexity: O(nlogn)
    // Space Complexity: O(1)
    // Sorting
    pub fn majority_element_3(nums: Vec<i32>) -> i32 {
        // Since we are assured that there will be a majority element which occurs more than nums.len() / 2 times, majority element will be at nums.len() / 2 index
        let mut nums = nums;
        nums.sort();
        nums[nums.len() / 2]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(Solution::majority_element(vec![3, 2, 3]), 3);
        assert_eq!(Solution::majority_element(vec![2, 2, 1, 1, 1, 2, 2]), 2);
        assert_eq!(Solution::majority_element(vec![1]), 1);
        assert_eq!(Solution::majority_element(vec![6, 5, 5]), 5);
    }
}
